script/LibStateDiffChecker.sol

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.15;
+
+import {VmSafe} from "forge-std/Vm.sol";
+import {stdJson} from "forge-std/StdJson.sol";
+import {console2 as console} from "forge-std/console2.sol";
+
+library LibStateDiffChecker {
+    VmSafe private constant vm = VmSafe(address(uint160(uint256(keccak256("hevm cheat code")))));
+
+    // The order of the fields in the struct is important, and must be alphabetical
+    // for foundry to properly decode data: https://book.getfoundry.sh/cheatcodes/parse-json#decoding-json-objects-into-solidity-structs
+    struct StorageDiffSpec {
+        address account;
+        bytes32 newValue;
+        bytes32 previousValue;
+        bytes32 slot;
+    }
+
+    struct StateDiffSpec {
+        uint256 chainId;
+        StorageDiffSpec[] storageSpecs;
+    }
+
+    error StateDiffMismatch(string field, bytes32 expected, bytes32 actual);
+
+    /// @notice Parses the JSON data to extract account access specifications.
+    /// @param _jsonData The JSON string containing the account access specifications.
+    /// @return stateDiffSpec_ An array of StorageDiffSpec structs parsed from the JSON data.
+    function parseDiffSpecs(string memory _jsonData) internal pure returns (StateDiffSpec memory stateDiffSpec_) {
+        stateDiffSpec_ = abi.decode(vm.parseJson(_jsonData), (StateDiffSpec));
+    }
+
+    /// @notice Filters account accesses to return only those that modified storage.
+    /// @param accountAccesses An array of account accesses recorded by the EVM.
+    /// @return stateDiffSpec_ An array of account accesses that modified storage.
+    function extractDiffSpecFromAccountAccesses(VmSafe.AccountAccess[] memory accountAccesses)
+        internal
+        pure
+        returns (StateDiffSpec memory stateDiffSpec_)
+    {
+        // Get the chain ID from the first account access.
+        stateDiffSpec_.chainId = accountAccesses[0].chainInfo.chainId;
+
+        // First iterate through the account accesses and storage accesses to count the number of storage modifications,
+        // so that we can allocate the correct amount of memory for the StorageDiffSpec.
+        uint256 modifiedCount = 0;
+        for (uint256 acctIndex = 0; acctIndex < accountAccesses.length; acctIndex++) {
+            for (
+                uint256 storageIndex = 0;
+                storageIndex < accountAccesses[acctIndex].storageAccesses.length;
+                storageIndex++
+            ) {
+                if (
+                    accountAccesses[acctIndex].storageAccesses[storageIndex].previousValue
+                        != accountAccesses[acctIndex].storageAccesses[storageIndex].newValue
+                ) {
+                    modifiedCount++;
+                }
```

Comment:
This will over-allocate length, since we might write different values to a slot multiple times? But that should be ok since later we only save the latest value

Reply:
They are not filtered out, but your test didn't change anything because `x` was declared as non-zero: 

https://github.com/ethereum-optimism/superchain-ops/blob/60f6c611d7d499f506ab46ceac4e320bced102bc/test/StateDiffChecker.t.sol#L11

Filtering in order to show only final state changes will require more work, and likely a change to how we do this tracking (ie. possibly using a mapping with `vm.dumpState`). 

But for now it is not supported.

Reply:
> your test didn't change anything because `x` was declared as non-zero:

Sorry what do you mean by this? That link gives me a 404, but L11 there on the latest commit is the `uint256 public x;` declaration

Reply:
I had done this in the state var declaration: 

```
contract Target {
   uint256 public x = 1;
...
```